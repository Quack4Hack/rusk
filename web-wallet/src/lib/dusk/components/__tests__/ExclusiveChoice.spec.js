import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";

import { ExclusiveChoice } from "..";

vi.mock("$lib/dusk/string", async (importOriginal) => {
  /** @type {typeof import("$lib/dusk/string")} */
  const original = await importOriginal();

  return {
    ...original,
    randomUUID: () => "some-generated-id",
  };
});

describe("ExclusiveChoice", () => {
  const stringOptions = ["one", "two", "three", "four"];

  /** @type {SelectOption[]} */
  const objectOptionsA = [
    { label: "one", value: "1" },
    { label: "two", value: "2" },
    { disabled: true, label: "three", value: "3" },
    { label: "four", value: "4" },
  ];

  /** @type {SelectOption[]} */
  const objectOptionsB = [
    { value: "1" },
    { value: "2" },
    { value: "3" },
    { value: "4" },
  ];

  const baseProps = {
    options: objectOptionsA,
    value: "2",
  };
  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  afterAll(() => {
    vi.doUnmock("$lib/dusk/string");
  });

  it("should render the `ExclusiveChoice` component", () => {
    const { container } = render(ExclusiveChoice, baseOptions);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should accept a custom name for the radio elements", () => {
    const props = {
      ...baseProps,
      name: "my-custom-name",
    };
    const { container } = render(ExclusiveChoice, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should accept an array of options object without labels and use the value as labels", () => {
    const props = {
      ...baseProps,
      options: objectOptionsB,
    };
    const { container } = render(ExclusiveChoice, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should accept an array of string as options", () => {
    const props = {
      ...baseProps,
      options: stringOptions,
    };
    const { container } = render(ExclusiveChoice, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should pass additional class names and attributes to the rendered element", () => {
    const props = {
      ...baseProps,
      className: "foo bar",
      id: "some-id",
    };
    const { container } = render(ExclusiveChoice, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should accept a change event handler", async () => {
    const changeHandler = vi.fn();
    const { component, container } = render(ExclusiveChoice, baseOptions);
    const target = /** @type {HTMLInputElement} */ (
      container.querySelector("input[value='4']")
    );

    component.$on("change", changeHandler);

    await fireEvent.click(target);

    expect(changeHandler).toHaveBeenCalledTimes(1);
    expect(changeHandler).toHaveBeenCalledWith(expect.any(Event));
    expect(target.checked).toBe(true);
  });
});
