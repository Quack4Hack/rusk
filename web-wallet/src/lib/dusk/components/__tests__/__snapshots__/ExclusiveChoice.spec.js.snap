// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExclusiveChoice > should accept a custom name for the radio elements 1`] = `
<div
  class="dusk-exclusive-choice"
  role="radiogroup"
>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-1"
    name="my-custom-name"
    type="radio"
    value="1"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-1"
  >
    one
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-2"
    name="my-custom-name"
    type="radio"
    value="2"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-2"
  >
    two
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    disabled=""
    id="some-generated-id-3"
    name="my-custom-name"
    type="radio"
    value="3"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-3"
  >
    three
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-4"
    name="my-custom-name"
    type="radio"
    value="4"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-4"
  >
    four
  </label>
</div>
`;

exports[`ExclusiveChoice > should accept an array of options object without labels and use the value as labels 1`] = `
<div
  class="dusk-exclusive-choice"
  role="radiogroup"
>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-1"
    name="some-generated-id"
    type="radio"
    value="1"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-1"
  >
    1
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-2"
    name="some-generated-id"
    type="radio"
    value="2"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-2"
  >
    2
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-3"
    name="some-generated-id"
    type="radio"
    value="3"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-3"
  >
    3
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-4"
    name="some-generated-id"
    type="radio"
    value="4"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-4"
  >
    4
  </label>
</div>
`;

exports[`ExclusiveChoice > should accept an array of string as options 1`] = `
<div
  class="dusk-exclusive-choice"
  role="radiogroup"
>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-one"
    name="some-generated-id"
    type="radio"
    value="one"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-one"
  >
    one
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-two"
    name="some-generated-id"
    type="radio"
    value="two"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-two"
  >
    two
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-three"
    name="some-generated-id"
    type="radio"
    value="three"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-three"
  >
    three
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-four"
    name="some-generated-id"
    type="radio"
    value="four"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-four"
  >
    four
  </label>
</div>
`;

exports[`ExclusiveChoice > should pass additional class names and attributes to the rendered element 1`] = `
<div
  class="dusk-exclusive-choice foo bar"
  id="some-id"
  role="radiogroup"
>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-1"
    name="some-generated-id"
    type="radio"
    value="1"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-1"
  >
    one
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-2"
    name="some-generated-id"
    type="radio"
    value="2"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-2"
  >
    two
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    disabled=""
    id="some-generated-id-3"
    name="some-generated-id"
    type="radio"
    value="3"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-3"
  >
    three
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-4"
    name="some-generated-id"
    type="radio"
    value="4"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-4"
  >
    four
  </label>
</div>
`;

exports[`ExclusiveChoice > should render the \`ExclusiveChoice\` component 1`] = `
<div
  class="dusk-exclusive-choice"
  role="radiogroup"
>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-1"
    name="some-generated-id"
    type="radio"
    value="1"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-1"
  >
    one
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-2"
    name="some-generated-id"
    type="radio"
    value="2"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-2"
  >
    two
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    disabled=""
    id="some-generated-id-3"
    name="some-generated-id"
    type="radio"
    value="3"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-3"
  >
    three
  </label>
  <input
    class="dusk-exclusive-choice__radio"
    id="some-generated-id-4"
    name="some-generated-id"
    type="radio"
    value="4"
  />
   
  <label
    class="dusk-exclusive-choice__label"
    for="some-generated-id-4"
  >
    four
  </label>
</div>
`;
